crates/rspack_regex/src/algo.rs

```diff
@@ -0,0 +1,99 @@
+use std::sync::Arc;
+
+use regex_syntax::hir::literal::ExtractKind;
+use regex_syntax::hir::{Hir, HirKind, Look};
+use rspack_error::{internal_error, Error};
+
+#[derive(Clone)]
+pub enum Algo {
+  FastCustom(Arc<dyn Fn(&str) -> bool + Send + Sync>),
+  Regress(regress::Regex),
+}
+
+impl std::fmt::Debug for Algo {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      Self::FastCustom(_) => write!(f, "FastCustom(...)"),
+      Self::Regress(arg0) => f.debug_tuple("Regress").field(arg0).finish(),
+    }
+  }
+}
+
+impl Algo {
+  pub fn new(expr: &str, flags: &str) -> Result<Algo, Error> {
+    if let Some(algo) = Self::try_fast_path(expr) {
+      Ok(algo)
+    } else {
+      regress::Regex::with_flags(expr, flags)
+        .map(Algo::Regress)
+        .map_err(|err| {
+          internal_error!("Can't construct regex `/{expr}/{flags}`, original error message: {err}")
+        })
+    }
+  }
+
+  pub fn try_fast_path(expr: &str) -> Option<Algo> {
+    let hir = regex_syntax::parse(expr).ok()?;
+    let seq = regex_syntax::hir::literal::Extractor::new()
+      .kind(ExtractKind::Suffix)
+      .extract(&hir);
```

Comment:
regress is also a parser right? It can't be used to detect this? 

Also, it's this just matching for an ending `$` inside the string?

Reply:
NO, `regress` is a regex engine, it can not provide `ast` or `hir` for regex

Reply:
`regex-syntax` parsed the original regular expression, and convert the pattern `****$` to a fast string search routine, this is what `regress` can't do.